### PR TITLE
Store credit reasons

### PR DESCRIPTION
**Pull request by jtapia**
#### Issue
- https://github.com/solidusio/solidus/issues/2110

#### What it does?
- Add migration to rename spree_store_credit_update_reasons to spree_store_credit_reasons
- Add store credit reason controller
- Add store credit reason model
- Add store credit reason views
- Update store credits controller to use store credit reason
- Add store credit routes and locales
- Update store credit model to use store credit reason
- Update store credit event model to use store credit reason
- Update store credit db seed
- Update factories to use store credit reason
- Update tests to use store credit reason

#### Preview
screen shot 2018-07-10 at 8 00 42 pm
screen shot 2018-07-10 at 8 10 16 pm
screen shot 2018-07-10 at 11 51 27 pm
screen shot 2018-07-10 at 11 51 47 pm
screen shot 2018-07-10 at 8 13 13 pm
screen shot 2018-07-10 at 8 14 19 pm
screen shot 2018-07-10 at 8 14 38 pm

### CODE REVIEW

File: core/app/models/spree/store_credit.rb [hunk 1]
```diff
@@ -43,7 +43,9 @@ class Spree::StoreCredit < Spree::PaymentSource
   before_destroy :validate_no_amount_used
   validate :validate_no_amount_used, if: :discarded?
 
-  attr_accessor :action, :action_amount, :action_originator, :action_authorization_code, :update_reason
+  belongs_to :store_credit_reason, class_name: 'Spree::StoreCreditReason', inverse_of: :store_credits
+
+  attr_accessor :action, :action_amount, :action_originator, :action_authorization_code, :store_credit_reason
```
Comment (kennyadsl) [inline core/app/models/spree/store_credit.rb]: I don't think `store_credit_reason` here is needed anymore

File: core/app/models/spree/store_credit_event.rb [hunk 2]
```diff
@@ -12,9 +12,9 @@ class StoreCreditEvent < Spree::Base
 
     belongs_to :store_credit
     belongs_to :originator, polymorphic: true
-    belongs_to :update_reason, class_name: "Spree::StoreCreditUpdateReason"
+    belongs_to :store_credit_reason, class_name: "Spree::StoreCreditReason"
 
-    validates_presence_of :update_reason, if: :action_requires_reason?
+    validates_presence_of :store_credit_reason_id, if: :action_requires_reason?
```
Comment (kennyadsl) [inline core/app/models/spree/store_credit_event.rb]: I think we can use `store_credit_reason` here

File: backend/app/controllers/spree/admin/store_credits_controller.rb [hunk 3]
```diff
@@ -78,18 +78,18 @@ def collection
         @collection = super.reverse_order
       end
 
-      def load_update_reasons
-        @update_reasons = Spree::StoreCreditUpdateReason.all.order(:name)
+      def load_reasons
+        @store_credit_reasons = Spree::StoreCreditReason.all.order(:name)
```
Comment (kennyadsl) [inline backend/app/controllers/spree/admin/store_credits_controller.rb]: maybe we should use the `active` scope here?

File: core/spec/models/spree/store_credit_event_spec.rb [hunk 4]
```diff
@@ -34,56 +34,56 @@
     end
   end
 
-  describe "update reason validation" do
+  describe "update store credit reason validation" do
     subject { event.valid? }
 
     context "adjustment event" do
-      context "has an update reason" do
+      context "has an store credit reason" do
```
Comment (kennyadsl) [inline core/spec/models/spree/store_credit_event_spec.rb]: `has an` -> `has a`

File: core/spec/models/spree/store_credit_event_spec.rb [hunk 5]
```diff
@@ -34,56 +34,56 @@
...
         let(:event) { build(:store_credit_adjustment_event) }
 
         it "returns true" do
           expect(subject).to eq true
         end
       end
 
-      context "doesn't have an update reason" do
-        let(:event) { build(:store_credit_adjustment_event, update_reason: nil) }
+      context "doesn't have an store credit reason" do
```
Comment (kennyadsl) [inline core/spec/models/spree/store_credit_event_spec.rb]: `doesn't have an` -> `doesn't have a`

File: core/spec/models/spree/store_credit_event_spec.rb [hunk 6]
```diff
@@ -34,56 +34,56 @@
...
           subject
-          expect(event.errors.full_messages).to match ["Update reason can't be blank"]
+          expect(event.errors.full_messages).to match ["Store credit reason can't be blank"]
         end
       end
     end
 
     context "invalidate event" do
-      context "has an update reason" do
+      context "has an store credit reason" do
```
Comment (kennyadsl) [inline core/spec/models/spree/store_credit_event_spec.rb]: same as above

File: core/spec/models/spree/store_credit_event_spec.rb [hunk 7]
```diff
@@ -34,56 +34,56 @@
...
         let(:event) { build(:store_credit_invalidate_event) }
 
         it "returns true" do
           expect(subject).to eq true
         end
       end
 
-      context "doesn't have an update reason" do
-        let(:event) { build(:store_credit_invalidate_event, update_reason: nil) }
+      context "doesn't have an store credit reason" do
```
Comment (kennyadsl) [inline core/spec/models/spree/store_credit_event_spec.rb]: same as above

File: core/spec/models/spree/store_credit_event_spec.rb [hunk 8]
```diff
@@ -34,56 +34,56 @@
...
+        it "adds an error message indicating the store credit reason is missing" do
           subject
-          expect(event.errors.full_messages).to match ["Update reason can't be blank"]
+          expect(event.errors.full_messages).to match ["Store credit reason can't be blank"]
         end
       end
     end
 
-    context "event doesn't require an update reason" do
+    context "event doesn't require an store credit reason" do
```
Comment (kennyadsl) [inline core/spec/models/spree/store_credit_event_spec.rb]: same as above

Comment (jtapia) [reply core/spec/models/spree/store_credit_event_spec.rb]: done

File: core/spec/models/spree/store_credit_event_spec.rb (hunk 7 above)
Comment (jtapia) [reply core/spec/models/spree/store_credit_event_spec.rb]: done

File: core/spec/models/spree/store_credit_event_spec.rb (hunk 6 above)
Comment (jtapia) [reply core/spec/models/spree/store_credit_event_spec.rb]: done

File: core/app/models/spree/store_credit.rb (hunk 1 above)
Comment (jtapia) [reply core/app/models/spree/store_credit.rb]: done

File: core/app/models/spree/store_credit_event.rb (hunk 2 above)
Comment (jtapia) [reply core/app/models/spree/store_credit_event.rb]: done

File: backend/app/controllers/spree/admin/store_credits_controller.rb (hunk 3 above)
Comment (jtapia) [reply backend/app/controllers/spree/admin/store_credits_controller.rb]: done

File: core/spec/models/spree/store_credit_event_spec.rb (hunk 4 above)
Comment (jtapia) [reply core/spec/models/spree/store_credit_event_spec.rb]: fixed

File: core/spec/models/spree/store_credit_event_spec.rb (hunk 5 above)
Comment (jtapia) [reply core/spec/models/spree/store_credit_event_spec.rb]: fixed

**Comment by kennyadsl** [comment]
Hey @jtapia, can you please rebase? I'd like to push merging this with the Core Team this week. Thanks!

**Comment by jtapia** [comment]
@kennyadsl done

**Pull request merged**